<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Ken B. Hanscombe" />

<meta name="date" content="2017-11-27" />

<title>Explore UK Biobank data</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">Explore UK Biobank data</h1>
<h4 class="author"><em>Ken B. Hanscombe</em></h4>
<h4 class="date"><em>2017-11-27</em></h4>



<p><br></p>
<p>After downloading and decrypting your UK Biobank (UKB) data with the supplied <a href="http://biobank.ctsu.ox.ac.uk/crystal/docs/UsingUKBData.pdf">UKB programs</a>, you have multiple files that need to be brought together to give you a tidy dataset to explore. The data file has column names that are the numeric field codes from the <a href="http://www.ukbiobank.ac.uk/about-data-showcase/">UKB data showcase</a>. <code>ukbtools</code> removes all the upfront data wrangling required to get a single dataset for statistical analysis, and provides tools to assist in quality control, query of disease diagnoses, and retrieval of genetic metadata.</p>
<p><br> <br></p>
<div id="preparing-a-ukb-fileset" class="section level2">
<h2>1. Preparing a UKB fileset</h2>
<p>Download<sup>ยง</sup> then decrypt your data and create a UKB fileset (.tab, .r, .html):</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="ex">ukb_unpack</span> ukbxxxx.enc key
<span class="ex">ukb_conv</span> ukbxxxx.enc_ukb r
<span class="ex">ukb_conv</span> ukbxxxx.enc_ukb docs</code></pre></div>
<p><code>ukb_unpack</code> decrypts your downloaded <code>ukbxxxx.enc</code> file, outputting a <code>ukbxxxx.enc_ukb</code> file. <code>ukb_conv</code> with the <code>r</code> flag converts the decrypted data to a tab-delimited file <code>ukbxxxx.tab</code> and an R script <code>ukbxxxx.r</code> that reads the tab file. The <code>docs</code> flag creates an html file containing a field-code-to-description table (among others).</p>
<p><br></p>
<p><sup>ยง</sup> Full details of the data download and decrypt process are given in the <a href="http://biobank.ctsu.ox.ac.uk/crystal/docs/UsingUKBData.pdf">Using UK Biobank Data</a> documentation .</p>
<p><br> <br></p>
</div>
<div id="making-a-dataset" class="section level2">
<h2>2. Making a dataset</h2>
<p>The function <code>ukb_df()</code> takes two arguments, the stem of your fileset and the path, and returns a dataframe with usable column names.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ukbtools)

my_ukb_data &lt;-<span class="st"> </span><span class="kw">ukb_df</span>(<span class="st">&quot;ukbxxxx&quot;</span>)</code></pre></div>
<p>You can also specify the path to your fileset if it is not in the current directory. For example, if your fileset is in a subdirectory of the working directory called <code>data</code></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">my_ukb_data &lt;-<span class="st"> </span><span class="kw">ukb_df</span>(<span class="st">&quot;ukbxxxx&quot;</span>, <span class="dt">path =</span> <span class="st">&quot;/full/path/to/my/fileset&quot;</span>)</code></pre></div>
<p>Use <code>ukb_df_field</code> to create a field-to-descriptive name key, as dataframe or named lookup vector.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">my_ukb_key &lt;-<span class="st"> </span><span class="kw">ukb_df_field</span>(<span class="st">&quot;ukbxxxx&quot;</span>, <span class="dt">path =</span> <span class="st">&quot;/full/path/to/my/fileset&quot;</span>)</code></pre></div>
<p><br></p>
<p><strong>Note:</strong> You can move the three files in your fileset after creating them with <code>ukb_conv</code>, but they should be kept together. <code>ukb_df()</code> automatically updates the read call in the R source file to point to the correct directory (the current directly by default, or the directory specified by <code>path</code>).</p>
<p><br></p>
<p><br></p>
<blockquote>
<p><strong>Memory and efficiency</strong></p>
<p>I highly recommend saving your new UKB dataset with <code>save(my_ukb_data, file = &quot;my_ukb_data.rda&quot;)</code>. You can load the data with <code>load(&quot;my_ukb_data.rda&quot;)</code>. Creating a UKB dataset from my largest UKB fileset which included a 2.6 GB .tab file, took approximately 8 mins; loading the same dataset after the <code>save</code> procedure, took about 20 seconds. The saved file was about 200 MB.</p>
</blockquote>
<p><br> <br></p>
<div id="multiple-downloads" class="section level3">
<h3>2.1 Multiple downloads</h3>
<p>If you have multiple UKB downloads, first read then merge them with your preferred method. <code>ukbtools</code> includes the function <code>ukb_df_full_join</code> which is a thin wrapper around <code>dplyr::full_join</code> applied recursively with <code>purrr::reduce</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ukbxxxx_data &lt;-<span class="st"> </span><span class="kw">ukb_df</span>(<span class="st">&quot;ukbxxxx&quot;</span>)
ukbyyyy_data &lt;-<span class="st"> </span><span class="kw">ukb_df</span>(<span class="st">&quot;ukbyyyy&quot;</span>)
ukbzzzz_data &lt;-<span class="st"> </span><span class="kw">ukb_df</span>(<span class="st">&quot;ukbzzzz&quot;</span>)

<span class="kw">ukb_df_full_join</span>(ukbxxxx_data, ukbyyyy_data, ukbzzzz_data)</code></pre></div>
<p><br></p>
<p><strong>Note:</strong> The join key is set to โeidโ only (default value of the <code>by</code> parameter). Any additional variables common to any two tables will have โ.xโ and โ.yโ appended to their names. If you are satisfied the additional variables are identical to the original, the copies can be safely deleted. For example, if <code>setequal(my_ukb_data$var, my_ukb_data$var.x)</code> is <code>TRUE</code>, then my_ukb_data$var.x can be dropped. A <code>dlyr::full_join</code> is like the set operation union in that all abservation from all tables are included, i.e., all samples are included even if they are not included in all datasets.</p>
<p><br></p>
<p>Repeated variable names <strong>within</strong> UKB datasets are unlikely to occur. <code>ukb_df</code> creates variable names by combining a snake_case descriptor with the variableโs <em>index</em> and <em>array</em>. This should be sufficient to uniquely identify the variable. However, if an <em>index_array</em> combination is incorrectly repeated in the original UKB data, this will result in a duplicated variable name. We observed two instances. The variables were encoded <em><field>โ0.0</em>, <em><field>โ1.0</em>, <em><field>โโ1.0</em>, and <code>ukb_df</code> created a variable named <em>var_0_0</em>, <em>var_1_0</em>, <em>var_1_0</em>. This is probably a typo that should have been <em><field>โ0.0</em>, <em><field>โ1.0</em>, <em><field>โ2.0</em>, consistent with UKB official documentation describing the field as having 3 values for index. We have provided <code>ukb_df_duplicated_names</code> to identify duplicated names within a dataset. This will allow the user to make changes as appropriate. We expect the occurrence of such duplicates will be rare.</p>
<p><br> <br></p>
</div>
</div>
<div id="exploring-primary-demographics-of-a-ukb-subset" class="section level2">
<h2>3. Exploring primary demographics of a UKB subset</h2>
<p>As an exploratory step you might want to look at the demographics of a particular subset of the UKB sample relative to a reference sample. For example, using the <code>nonmiss.var</code> argument of <code>ukb_context</code> will produce a plot of the primary demographics (sex, age, ethnicity, and Townsend deprivation score) and employment status and assessment centre, for the subsample with data on your variable of interest compared to those without data (i.e. <code>NA</code>).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ukb_context</span>(my_ukb_data, <span class="dt">nonmiss.var =</span> <span class="st">&quot;my_variable_of_interest&quot;</span>)</code></pre></div>
<p><br></p>
<p>It is also possible to supply a logical vector with <code>subset.var</code> to define the subset and reference sample. This is particularly useful for understanding a subgroup within the UKB study, e.g., obese individuals below age 50.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">subgroup_of_interest &lt;-<span class="st"> </span>(my_ukb_data<span class="op">$</span>bmi <span class="op">&gt;</span><span class="st"> </span><span class="dv">40</span> <span class="op">&amp;</span><span class="st"> </span>my_ukb_data<span class="op">$</span>age <span class="op">&lt;</span><span class="st"> </span><span class="dv">50</span>) 
<span class="kw">ukb_context</span>(my_ukb_data, <span class="dt">subset.var =</span> subgroup_of_interest)</code></pre></div>
<p><br> <br></p>
</div>
<div id="retrieving-icd-diagnoses" class="section level2">
<h2>4. Retrieving ICD diagnoses</h2>
<p>All ICD related functions begin <code>ukb_icd_</code>. Type <code>ukb_icd_</code> tab to see the family of functions. The full ICD โcode-meaningโ tables are available as datasets (<code>icd9codes</code>, <code>icd10codes</code>). ICD chapter-level tables describing disease blocks are also available for query as datasets (<code>icd9chapters</code>, <code>icd10chapters</code>)</p>
<p><br></p>
<p>To retrieve the full diagnosis of an individual (combine multiple individuals with <code>c()</code>)</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ukb_icd_diagnosis</span>(my_ukb_data, <span class="dt">id =</span> <span class="st">&quot;0000000&quot;</span>, <span class="dt">icd.version =</span> <span class="dv">10</span>)</code></pre></div>
<p><br></p>
<p>To retrieve the โmeaningโ of an ICD code use <code>icd_code</code>. Again, you can look up multiple codes by combining them with <code>c()</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ukb_icd_code_meaning</span>(<span class="dt">icd.code =</span> <span class="st">&quot;I74&quot;</span>, <span class="dt">icd.version =</span> <span class="dv">10</span>)</code></pre></div>
<p><br></p>
<p>Search for a class of diseases with a keyword. Supplying multiple keywords with <code>c()</code>, will return all ICD entries containing <em>any</em> of the keywords.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ukb_icd_keyword</span>(<span class="st">&quot;cardio&quot;</span>, <span class="dt">icd.version =</span> <span class="dv">10</span>)</code></pre></div>
<p><br></p>
<p>You can calculate the prevalence of a diagnosis in the UKB study (or a subset of the full sample) using <code>ukb_icd_prevalence</code>. The <code>icd.diagnosis</code> argument takes a regular expression, and so can also be used to retrieve prevalence of a disease โclassโ, i.e., the proportion of individuals with <em>any</em> diagnosis in the disease class.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># ICD-10 code I74, Arterial embolism and thrombosis</span>
<span class="kw">ukb_icd_prevalence</span>(my_ukb_data, <span class="dt">icd.version =</span> <span class="dv">10</span>, <span class="dt">icd.diagnosis =</span> <span class="st">&quot;I74&quot;</span>)

<span class="co"># ICD-10 chapter 9, disease block I00โI99, Diseases of the circulatory system</span>
<span class="kw">ukb_icd_prevalence</span>(my_ukb_data, <span class="dt">icd.version =</span> <span class="dv">10</span>, <span class="dt">icd.diagnosis =</span> <span class="st">&quot;I&quot;</span>)

<span class="co"># ICD-10 chapter 2, C00-D49, Neoplasms </span>
<span class="kw">ukb_icd_prevalence</span>(my_ukb_data, <span class="dt">icd.version =</span> <span class="dv">10</span>, <span class="dt">icd.diagnosis =</span> <span class="st">&quot;C|D[0-4].&quot;</span>)</code></pre></div>
<p><br></p>
<p>To retrieve frequency for one or more ICD diagnoses by the levels of a reference variable, e.g., sex (male or female) use <code>ukb_icd_freq_by</code>. If the variable is continuous, it is divided into N approximately equal-sized groups (default = 10) within which ICD diagnosis frequency is calculated. ukb_icd_freq_by also includes an option to produce a figure of ICD diagnosis frequency by reference variable. Diagnoses of interest are passed to <code>icd.code</code>. The default ICD codes are the WHO top 3 cause of death worldwide (2015): coronary artery disease (CAD), cerebrovascular disease/ stroke, lower respiratory tract infection (LTRI).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># To plot the frequency of the default ICD codes with respect to BMI</span>
<span class="kw">ukb_icd_freq_by</span>(my_ukb_data, <span class="dt">reference.var =</span> <span class="st">&quot;body_mass_index_bmi_0_0&quot;</span>, <span class="dt">freq.plot =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p><br></p>
<p><strong>Note.</strong> Setting <code>freq.plot = FALSE</code> (default) returns a dataframe of the frequencies. Values for the reference variable group ranges are in the column โgroupโ.</p>
<p><br> <br></p>
</div>
<div id="retrieving-genetic-metadata" class="section level2">
<h2>5. Retrieving genetic metadata</h2>
<p>If you are doing any downstream genetic analyses, you will need the genetic metadata (which should be in you phenotype dataset). Detailed information is available on UKB <a href="http://www.ukbiobank.ac.uk/wp-content/uploads/2014/04/UKBiobank_genotyping_QC_documentation-web.pdf">genotyping and quality control</a> and <a href="http://www.ukbiobank.ac.uk/wp-content/uploads/2014/04/imputation_documentation_May2015.pdf">imputation and association</a>.</p>
<p>All genetic metadata related functions begin <code>ukb_gen_</code>. Typing <code>ukb_gen_</code> tab opens a dropdown menu for this group of functions.</p>
<p><br> <br></p>
<p>You can collect the genetic metadata (including recommended exclusions, genetic sex, genetic ethnicity, chip, etc.), and principal components with</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">my_gen_meta &lt;-<span class="st"> </span><span class="kw">ukb_gen_meta</span>(my_ukb_data)
my_gen_pcs &lt;-<span class="st"> </span><span class="kw">ukb_gen_pcs</span>(my_ukb_data)</code></pre></div>
<p><br> <br></p>
<p>A list of IDs for recommended exclusions and heterozygosity outliers (+/- 3*SD) can be retrieved</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ukb_gen_excl</span>(my_ukb_data)
<span class="kw">ukb_gen_het</span>(my_ukb_data)</code></pre></div>
<p>For a <code>data.frame</code> of raw heterozygosity data</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ukb_gen_het</span>(my_ukb_data, <span class="dt">all.het =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p><br> <br></p>
<p><code>ukb_gen_rel</code> returns a <code>data.frame</code> with <code>id</code>, <code>pair</code> (a numeric identifier for related pairs), and <code>kinship</code> (kinship coefficient). For a count of related samples by degree of relatedness use <code>ukb_gen_rel_count</code>. Set the argument <code>plot = TRUE</code> to replicate the plot on page 15 of the <a href="http://www.ukbiobank.ac.uk/wp-content/uploads/2014/04/UKBiobank_genotyping_QC_documentation-web.pdf">UKB genotyping and quality control documentation</a>, for your subgroup of interest.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">my_gen_rel &lt;-<span class="st"> </span><span class="kw">ukb_gen_rel</span>(my_ukb_data)

<span class="co"># To get a count and plot of degree of relatedness</span>
<span class="kw">ukb_gen_rel_count</span>(my_gen_rel, <span class="dt">plot =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p><br> <br></p>
</div>
<div id="read-and-write" class="section level2">
<h2>6. Read and write</h2>
<p><code>ukbtools</code> includes functions to write phenotype and covariate files for <a href="https://jmarchini.org/bgenie/">BGENIE</a> and <a href="https://www.cog-genomics.org/plink2">PLINK</a>. <strong>BGENIE</strong> phenotype and covariate files are space-delimited, include column names, and have missing values coded as -999. They must also be in .sample file order. <code>ukb_gen_write_bgenie</code> sorts input data to match .sample file id order and writes the data to disk.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Read .sample file supplied with bulk genetic data</span>
my_sample_file &lt;-<span class="st"> </span><span class="kw">ukb_gen_read_sample</span>(<span class="st">&quot;path/to/sample_file&quot;</span>)

<span class="co"># Write a BGENIE format phenotype or covariate file</span>
<span class="kw">ukb_gen_write_bgenie</span>(
  my_ukb_data, 
  <span class="dt">path =</span> <span class="st">&quot;path/to/bgenie_input_file&quot;</span>, 
  <span class="dt">ukb.sample =</span> my_sample_file, 
  <span class="dt">ukb.variables =</span> <span class="kw">c</span>(<span class="st">&quot;variable1&quot;</span>, <span class="st">&quot;variable2&quot;</span>, <span class="st">&quot;variable3&quot;</span>)
)</code></pre></div>
<p><br></p>
<p><strong>Note.</strong> The <a href="https://jmarchini.org/bgenie-usage/">BGENIE usage page</a> uses the example files example.pheno and example.cov - it is not clear whether the suffixes are obligatory. Use them to be safe.</p>
<p><br> <br></p>
<p><strong>PLINK</strong> phenotype and covariate files are either space- or tab-delimited, column names are optional, first two columns must contain family ID and individual ID respectively, and missing values are โnormally expected to be encoded as -9โ but also โnonnumeric values such as โNAโ โฆ (are) treated as missingโ. <code>ukb_gen_write_plink</code> writes a space-delimited file with column names, UKB ID is automatically written to column 1 and 2 and labelled FID IID, and missing values are coded as <code>NA</code>. The missing value to be used can be changed with the <code>na.strings</code> argument. See <a href="https://www.cog-genomics.org/plink/1.9/input#pheno">PLINK standard data input</a> for further details.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Write a PLINK format phenotype or covariate file</span>
<span class="kw">ukb_gen_write_plink</span>(
  my_ukb_data, 
  <span class="dt">path =</span> <span class="st">&quot;path/to/plink_input_file&quot;</span>, 
  <span class="dt">ukb.variables =</span> <span class="kw">c</span>(<span class="st">&quot;variable1&quot;</span>, <span class="st">&quot;variable2&quot;</span>, <span class="st">&quot;variable3&quot;</span>)
)</code></pre></div>
<p><strong>PLINK</strong> does not require that individuals in phenotype and covariate files are in any particular order, but you may want to reconcile the individuals you include in your analysis with those in the fam file (from your hard-called data). Read the fam file into R with <code>ukb_gen_read_fam</code></p>
<p><br> <br></p>
<div id="exclusions" class="section level3">
<h3>6.1 Exclusions</h3>
<p><strong>Note.</strong> The exclusions referred to in this section are the combined UKB recommended exclusions, Affymetrix quality control for samples, bileve genotype quality control, heterozygosity outliers (+/- 3*SD), genetic ethnicity outliers (based on Genetic Ethnic Grouping, field code 22006), and one random member of each related pair (including Duplicates/MZ twins, 1st-, 2nd-, and 3rd-degree relatives).</p>
<p><br></p>
<p><strong>BGENIE</strong> does not have an option to read exclusions. You can replace data values in a phenotype with <code>NA</code> where the individual is to-be-excluded based on genetic metadata considerations. Writing the updated variable to your phenotype file (with the supplied write functions), effectively excludes the individuals from any analysis.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">my_ukb_data<span class="op">$</span>height_excl_na &lt;-<span class="st"> </span><span class="kw">ukb_gen_excl_to_na</span>(my_ukb_data, <span class="dt">x =</span> <span class="st">&quot;height&quot;</span>)</code></pre></div>
<p><br></p>
<p><strong>PLINK</strong> <code>--remove</code> takes a space- or tab-delimited file with family IDs in the first column and individual IDs in the second column, without column names. See <a href="https://www.cog-genomics.org/plink/1.9/filter#indiv">PLINK input filtering</a> for further details. (The missing value approach described above also works for PLINK.)</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ukb_gen_write_plink_excl</span>(<span class="st">&quot;path/to/plink_input_file&quot;</span>)</code></pre></div>
</div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
